### PR TITLE
Project voting

### PR DESCRIPTION
**Pull request by JordoBrooks**
- Developers can create a TimelineProposal to initiate a vote among contributors to replace existing timeline with pendingTimeline.

- Developers can create a MilestoneCompletionSubmission that initiates a vote of approval among contributors to accept the current milestone as complete and release the funds for the next milestone.

- When voting is active (for either a TimelineProposal or a MilestoneCompletionSubmission), contributors to the project are eligible to place a vote of approval, or a vote of disapproval. Developer can choose to complete the voting with success dependent on one of two conditions:

1. 2 weeks have passed and over 75% approval acheived among those who voted.
2. Over 75% of the total contributors have voted to approve.

### CODE REVIEW

File: smart-contracts/ethereum/contracts/Project.sol [hunk 1]
```diff
@@ -67,68 +101,222 @@ contract Project {
             isComplete: false
             });
 
-        pendingTimeline.push(newMilestone);
+        if (_isPending) {
+            // There must not be an active timeline proposal
+            require(!timelineProposal.isActive);
+            // There must be an active timeline already
+            require(timeline.isActive);
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/Project.sol]: Can we send along error messages now? (if not, can we still attempt it here until it does kick in)

File: smart-contracts/ethereum/contracts/Project.sol [hunk 2]
```diff
@@ -67,68 +101,222 @@ contract Project {
...
+        // TimelineProposal must be active
+        require(timelineProposal.isActive == true);
+
+        FundingService fs = FundingService(fundingService);
+        uint numContributors = fs.getProjectContributorList(this).length;
+        uint numVoters = timelineProposal.approvalCount + timelineProposal.disapprovalCount;
+        bool isTwoWeeksLater = now >= timelineProposal.timestamp + 2 weeks;
+        uint votingThreshold = numVoters * 75 / 100;
+
+        // Proposal needs >75% total approval, or for 2 days to have passed and >75% approval among voters
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/Project.sol]: I think it means to say 2 weeks. Btw critical paths/business logic like this needs a couple tests for sure.

File: smart-contracts/ethereum/contracts/Project.sol [hunk 3]
```diff
@@ -67,68 +101,222 @@ contract Project {
...
+        timeline.milestones[activeMilestoneIndex].isComplete = true;
+        activeMilestoneIndex++;
+
+        // todo - make funds available for developer to withdraw
+
+        delete(milestoneCompletionSubmission);
         delete(pendingTimeline);
+
+        // Push completed milestone
+        pendingTimeline.milestones.push(timeline.milestones[activeMilestoneIndex - 1]);
```
Comment (ericmuyser) [inline smart-contracts/ethereum/contracts/Project.sol]: Why?

File: smart-contracts/ethereum/contracts/Project.sol (hunk 2 above)
Comment (ericmuyser) [reply smart-contracts/ethereum/contracts/Project.sol]: I found the tests! Looks great!

**Comment by ericmuyser** [comment]
Very minor comments before merge. Great work! 🚀

File: smart-contracts/ethereum/contracts/Project.sol (hunk 1 above)
Comment (JordoBrooks) [reply smart-contracts/ethereum/contracts/Project.sol]: Good call. Probably best to add error messages to all the require statements in our code base.

**Comment by ericmuyser** [comment]
Could you write a test for this https://github.com/hyperbridge/funding-protocol/pull/14/commits/3c3ac7491ecbab016fb94b92ddbd13455e3a4068 ?

**Comment by JordoBrooks** [comment]
Made some last minute additions to the PR (https://github.com/hyperbridge/funding-protocol/pull/14/commits/7c4abc4574988352c1968a3133f56a73787ff39b). In particular, I realized that pendingTimelines (and by extension proposals for new timelines) should take into account the milestones which have already been completed/released funds.

**Pull request merged**